File: architecture-skills/design-patterns/examples/template-method/template-method.swift
// Template Method pattern in Swift.
//
// Swift's idiomatic form is a protocol extension: the extension's
// default method is the template and the protocol requirements are the
// varying steps — no base class needed, and it works for structs.
// Composition (passing the steps as closures) is often preferred over
// any inheritance-based form.

import Foundation

struct LineItem {
    let product: String
    let units: Int
    let price: Double

    var subtotal: Double { Double(units) * price }
}

protocol SalesReport {
    func header() -> String
    func row(_ item: LineItem) -> String
    func footer(total: Double) -> String
}

extension SalesReport {
    /// The template method: iteration and totalling never vary.
    func generate(_ items: [LineItem]) -> String {
        var lines = [header()]
        var total = 0.0
        for item in items {
            lines.append(row(item))
            total += item.subtotal
        }
        lines.append(footer(total: total))
        return lines.joined(separator: "\n")
    }
}

struct CsvReport: SalesReport {
    func header() -> String {
        "product,units,price,subtotal"
    }

    func row(_ item: LineItem) -> String {
        String(format: "%@,%d,%.2f,%.2f", item.product, item.units, item.price, item.subtotal)
    }

    func footer(total: Double) -> String {
        String(format: "total,,,%.2f", total)
    }
}

struct MarkdownReport: SalesReport {
    func header() -> String {
        "| Product | Units | Subtotal |\n| --- | --- | --- |"
    }

    func row(_ item: LineItem) -> String {
        String(format: "| %@ | %d | $%.2f |", item.product, item.units, item.subtotal)
    }

    func footer(total: Double) -> String {
        String(format: "\n**Total: $%.2f**", total)
    }
}

let items = [
    LineItem(product: "Keyboard", units: 2, price: 80.00),
    LineItem(product: "Mouse", units: 1, price: 25.50),
    LineItem(product: "Monitor", units: 3, price: 210.00),
]

let reports: [(String, SalesReport)] = [
    ("CsvReport", CsvReport()),
    ("MarkdownReport", MarkdownReport()),
]

for (name, report) in reports {
    print("--- \(name) ---")
    print(report.generate(items))
    print()
}

File: architecture-skills/design-patterns/examples/template-method/template-method.py
"""Template Method pattern in Python.

The base class fixes the report skeleton (generate); subclasses supply
the formatting steps. Note: Python often prefers composition or a
higher-order function taking the steps as arguments over inheritance;
the ABC form shown here is the classic pattern.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass


@dataclass(frozen=True)
class LineItem:
    product: str
    units: int
    price: float

    @property
    def subtotal(self) -> float:
        return self.units * self.price


class SalesReport(ABC):
    def generate(self, items: list[LineItem]) -> str:
        """The template method: the skeleton is fixed here.

        Iteration and totalling never vary; only formatting does.
        """
        lines = [self.header()]
        lines.extend(self.row(item) for item in items)
        lines.append(self.footer(sum(item.subtotal for item in items)))
        return "\n".join(lines)

    @abstractmethod
    def header(self) -> str: ...

    @abstractmethod
    def row(self, item: LineItem) -> str: ...

    @abstractmethod
    def footer(self, total: float) -> str: ...


class CsvReport(SalesReport):
    def header(self) -> str:
        return "product,units,price,subtotal"

    def row(self, item: LineItem) -> str:
        return f"{item.product},{item.units},{item.price:.2f},{item.subtotal:.2f}"

    def footer(self, total: float) -> str:
        return f"total,,,{total:.2f}"


class MarkdownReport(SalesReport):
    def header(self) -> str:
        return "| Product | Units | Subtotal |\n| --- | --- | --- |"

    def row(self, item: LineItem) -> str:
        return f"| {item.product} | {item.units} | ${item.subtotal:.2f} |"

    def footer(self, total: float) -> str:
        return f"\n**Total: ${total:.2f}**"


if __name__ == "__main__":
    items = [
        LineItem("Keyboard", 2, 80.00),
        LineItem("Mouse", 1, 25.50),
        LineItem("Monitor", 3, 210.00),
    ]
    for report in (CsvReport(), MarkdownReport()):
        print(f"--- {type(report).__name__} ---")
        print(report.generate(items))
        print()
